### ISSUE TITLE
[Bug]: Fyne doesn't compile on termux.

### ISSUE THREAD

**Issue by DuilioPerez** [fyne-io/fyne#5173]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I've followed the instructions from your [getting start page](https://docs.fyne.io/started/) to install fyne in Termux, but I've found some issues:
1 - It can't compile programs directly for Termux:
2 - Android API level used is too low:

### How to reproduce

1: Run `go build` in the firs example app
2: Install an app

### Screenshots

1:
![Screenshot_2024-09-29_14-30-59](https://github.com/user-attachments/assets/e9f3f93c-2041-4776-9ad9-c34ed7a517f7)

2:
![Screenshot_20240929-143749_WhatsApp](https://github.com/user-attachments/assets/3f609c2a-7d33-4c39-ae12-7d80e4fb7c63)

### Example code

```go
package main
import ( 
  "fyne.io/fyne/v2/app"
  "fyne.io/fyne/v2/widget"
)

func main() {
  a := app.New()
  w := a.NewWindow("Hello World")
  w.SetContent(widget.NewLabel("Hello World!"))
  w.ShowAndRun()
}
```

### Fyne version

fyne cli version: (devel)

### Go compiler version

go version go1.23.1 android/arm64

### Operating system and version

Android 11

**Comment by andydotxyz** [comment]
> Fyne version
>
> fyne cli version: (devel)

Please list the library version not the cli.

More info:

Please share what is at line 136 of `/usr/include/EGL/egl.h` - the specification states that it is implementation specific but obviously for many platforms it is working, so we need to see what yours is expecting.

**Comment by andydotxyz** [comment]
Correcting title as this is compile specific. If you compiled the android app elsewhere and copied it on then it would execute as far as I can see.

**Comment by andydotxyz** [comment]
Oh, hold on - isn't termux a linux emulator for Android? In that case you should probably try "fyne package -os linux" because it's trying to build an android app from a linux terminal.

**Comment by DuilioPerez** [comment]
> Oh, hold on - isn't termux a linux emulator for Android? In that case you should probably try "fyne package -os linux" because it's trying to build an android app from a linux terminal.

That command have the same error. Termux is an Android Linux emulator, but it uses bionic instead of glibc. It's default compiles for linux-android.

The content of line 136 is:
`EGLAPI EGLSurface EGLAPIENTRY eglCreateWindowSurface (EGLDisplay dpy, EGLConfig config, EGLNativeWindowType win, const EGLint *attrib_list);`

**Comment by DuilioPerez** [comment]
And the other issue is about Android API Level too low

**Comment by DuilioPerez** [comment]
Fyne version is 2.5.1

**Comment by andydotxyz** [comment]
> The content of line 136 is:
> EGLAPI EGLSurface EGLAPIENTRY eglCreateWindowSurface (EGLDisplay dpy, EGLConfig config, EGLNativeWindowType win, const EGLint *attrib_list);

Thanks, and what is the definition of `EGLNativeWindowType` on that platform (also, where is it defined?)

> And the other issue is about Android API Level too low

Google Play Protect hates our backwards compatibility - it is not a problem, just a nuisance.

**Comment by DuilioPerez** [comment]
There is a related issue, maybe that information can help:
https://github.com/hajimehoshi/ebiten/issues/3057

**Comment by DuilioPerez** [comment]
And with the issue related to API Level, it could be usefull to have a way to use a API Level which allow external distributions. Maybe an option on the CLI tool.

**Comment by andydotxyz** [comment]
> it could be usefull to have a way to use a API Level which allow external distributions. 

It is not clear what distribution is blocked by our backwards compatibility.

**Comment by DuilioPerez** [comment]
Well, it seems google play blocks apps according to the used target sdk. So isn't the minimun API Level what needs to be updated but the target sdk, acording to the [Android SDK documentation](https://developer.android.com/google/play/requirements/target-sdk), which is updated every year, so that's the only which need to be updated.

**Comment by andydotxyz** [comment]
It is not blocked - it is a warning that you can accept.

Our target SDK is above the minimum required. If it wasn't then Google would have removed the listing from the store.

**Comment by andydotxyz** [comment]
> That command have the same error. Termux is an Android Linux emulator, but it uses bionic instead of glibc. It's default compiles for linux-android.

Then please try setting GOOS=linux. I can tell it is compiling for
Android because it is linking to the mobile driver...

**Comment by andydotxyz** [comment]
Actually - from reading other bug reports it seems like `fyne package -os android` is what you should be trying.

**Comment by DuilioPerez** [comment]
Yes, but that will be create an APK. Here I would like to create a desktop application for Termux since it handles its own binaries, but it seems fyne isn't compatible yet. This could be related to the Golang package available in the repositories of Termux, since other GUI libraries for other languages works fines, like SDL2 and raylib for C; SFML, Qt and WxWidgets for C++, TkInter for python and many others.  I think we should close this issue since it's hard to give a solution here.

**Comment by andydotxyz** [comment]
You'll need to help us understand - are these supposed to be linux apps or Android?
The bug opened with a crash because the Android implementation is not compatible.

But if you're looking to build a Linux app then it's a red-herring and you should follow the descriptions above for building a Linux app instead.

**Comment by DuilioPerez** [comment]
Creating an android app using fyne works! The app once built can be installed, it gives the warning of targetting an old Android API, but it works. But if I want to compile the same app for launching inside Termux, like a desktop app, it have some conflict that prevents compilation. I guess is because it, even explicitly setting the OS as Linux, it detect it's Android. That's a common error in multiplatform APIs, I said because I compiled SDL3 and I removed the line that detected Android. It seems there are other ways, but I don't have enough knowledge about Go to give a solution. In C, there's a macro called Termux which indicates we're compiling using the compiler provided by the distribution, and not the compiler from the Android NDK.

**Comment by DuilioPerez** [comment]
As the warning about NDK isn't an actual bug, you can close this issue. And I've saw a bug report in the Termux GitHub repositories about graphics in Go, so I guess is a issue with the distribution of Go for that platform, and that's the reason why I say it won be fixed here. I've didn't work with fyne since this bug report, so I don't know if that issue was fixed, but if nor, it's highly probable it will be fixed in the future.

**Comment by andydotxyz** [comment]
I'm glad the Android build worked.

The errors you posted all related to Android compile. As far as I can see there there was no feedback to testing `fyne package -os linux` or `GOOS=linux go build` - either of those will override the detected platform.

**Comment by DuilioPerez** [comment]
> I'm glad the Android build worked.
> 
> The errors you posted all related to Android compile. As far as I can see there there was no feedback to testing `fyne package -os linux` or `GOOS=linux go build` - either of those will override the detected platform.

Well, I tried that, but it didn't work. I guess is a bug in the Go package in Termux repositories. It uses the android driver even specifying Linux, in which case should use X11 for graphics

**Comment by andydotxyz** [comment]
How it connects to graphics has nothing to do with the Go compiler, that is all Fyne.

It would be helpful if you shared the error.

**Comment by DuilioPerez** [comment]
I've reinstalled fyne and tried. It don't work:
```bash
~/helloWorld $ GOOS=linux go build main.go
package command-line-arguments
        imports fyne.io/fyne/v2/app
        imports fyne.io/fyne/v2/internal/driver/glfw
        imports fyne.io/fyne/v2/internal/driver/common
        imports fyne.io/fyne/v2/internal/painter/gl
        imports github.com/go-gl/gl/v3.1/gles2: build constraints exclude all Go files in /data/data/com.termux/files/home/go/pkg/mod/github.com/go-gl/gl@v0.0.0-20211210172815-726fda9656d6/v3.1/gles2
~/helloWorld $ go build main.go
# fyne.io/fyne/v2/internal/driver/mobile/app
android.c:203:52: error: incompatible pointer to integer conversion passing 'ANativeWindow *' (aka 'struct ANativeWindow *') to parameter of type 'EGLNativeWindowType' (aka 'unsigned long') [-Wint-conversion]
../../usr/include/EGL/egl.h:136:109: note: passing argument to parameter 'win' here
~/helloWorld $ fyne package -os linux -name helloWorld
# net
../../usr/lib/go/src/net/cgo_resnew.go:20:76: cannot use _Ctype_socklen_t(len(b)) (value of type _Ctype_socklen_t) as _Ctype_size_t value in argument to (_C2func_getnameinfo)

error building application: exit status 1
~/helloWorld $ GOOS=linux go mod tidy
~/helloWorld $
~/helloWorld $ GOOS=linux go build main.go
package command-line-arguments
        imports fyne.io/fyne/v2/app
        imports fyne.io/fyne/v2/internal/driver/glfw
        imports fyne.io/fyne/v2/internal/driver/common
        imports fyne.io/fyne/v2/internal/painter/gl
        imports github.com/go-gl/gl/v3.1/gles2: build constraints exclude all Go files in /data/data/com.termux/files/home/go/pkg/mod/github.com/go-gl/gl@v0.0.0-20211210172815-726fda9656d6/v3.1/gles2
~/helloWorld $
```
